connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -183,6 +183,60 @@ message ExecutePlanResponse {
   }
 }
 
+// Request to transfer client-local artifacts.
+message AddArtifactsRequest {
+
+  // Definition of an Artifact.
+  message Artifact {
+    // The name of the artifact is expected in the form of a "Relative Path" that is made up of a
+    // sequence of directories and the final file element.
+    // Examples of "Relative Path"s: "jars/test.jar", "classes/xyz.class", "abc.xyz", "a/b/X.jar".
+    // The server is expected to maintain the hierarchy of files as defined by their name. (i.e
+    // The relative path of the file on the server's filesystem will be the same as the name of
+    // the provided artifact)
+    string name = 1;
+    // Raw data.
+    bytes data = 2;
+  }
+
+  // A number of small artifacts batched into a single RPC.
+  message Batch {
+    repeated Artifact artifacts = 1;
+  }
+
+  // The client_id is set by the client to be able to collate streaming responses from
+  // different queries.
+  string client_id = 1;
+
+  // User context
+  UserContext user_context = 2;
+
+  // The payload is either a batch of artifacts or a partial chunk of a large artifact.
+  oneof payload {
+    Batch batch = 3;
+    // A large artifact chunked into multiple requests. The server side should assume that the
+    // artifact has been completely uploaded either when it encounters a new artifact name, or
+    // when the the stream is completed.
+    Artifact chunk = 4;
```

Comment:
This protocol basically defines the end of an artifact by the start of a new artifact. It feels awkward. Why not include some metadata, such as total expected length for an artifact, to make it easy for the server to determine up front whether it is going to receive a full artifact or just a chunk of it? This could also make it easier for the server-side to, for example, allocate the entire required space up front.

Reply:
Thanks for the feedback.

I've introduced two new message types to handle this:
The `BeginChunkedArtifact` message now contains the metadata of the chunked artifact as well as the initial chunk. Messages following this would be of `ArtifactChunk` type which represents the difference between an artifact chunk and  a "batch" of artifact (which may either be a single, complete artifact or multiple complete artifacts).

Reply:
This makes sense to me. Do we also need `EndChunkedArtifact`? The server-side can infer it based on `num_chunks`, but I wonder if being explicit is better?

Reply:
> Do we also need EndChunkedArtifact? 

The `BeginChunkedArtifact` contains some metadata but I'm not sure what other data we can include in `EndChunkedArtifact` that's also not available in `BeginChunkedArtifact` so I've left it out in order to keep message types minimal. However, no strong preference for either way.